src/FrameId.h

```diff
@@ -2,29 +2,36 @@
 #define FRAMEID_H_INCLUDED
 
 #include <cstddef>
+#include <cstdint>
+#include <exception>
+#include <limits>
 
 struct FrameId {
-	static constexpr int Invalid = -1;
+	static constexpr uint32_t Invalid = std::numeric_limits<uint32_t>::max();
 	constexpr FrameId() :
 		m_id(Invalid) {}
-	constexpr FrameId(int new_id) :
+	constexpr FrameId(uint32_t new_id) :
 		m_id(new_id) {}
 
-	constexpr operator bool() const { return m_id > Invalid; }
-	constexpr operator int() const { return m_id; }
+	constexpr operator bool() const { return m_id != Invalid; }
+	constexpr operator uint32_t() const { return m_id; }
 	constexpr operator size_t() const { return m_id; }
 
 	constexpr bool operator==(FrameId rhs) const { return m_id == rhs.m_id; }
 	constexpr bool operator!=(FrameId rhs) const { return m_id != rhs.m_id; }
 
-	constexpr bool valid() const { return m_id > Invalid; }
-	constexpr int id() const { return m_id; }
+	constexpr bool valid() const { return m_id != Invalid; }
+	constexpr size_t id() const {
+		// Cannot use an `if` in constexpr context in C++11, sorry
+		return valid() ? static_cast<size_t>(m_id) : (std::terminate(), 0);
```

Comment:
Apologies if this wasn't clear - unconditionally aborting is an anti-pattern we would prefer to remove from the code. This function should simply return the `m_id` value converted to `size_t` - if we care if it's invalid, calling code is responsible for checking the value instead of the value accessor itself. Additionally, (with an implicit type promotion) the value returned by `id()` can be compared with the `FrameId::Invalid` value to determine if it's invalid with no issues.

Reply:
Ouch. Sorry, my bad. I totally misunderstood this point and the one related to `lua_Unsigned`.